daemon/containerd/image_delete.go

```diff
@@ -194,9 +199,71 @@ func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, force,
 		}
 	}
 
+	if len(options.Platforms) > 0 {
+		return i.deleteImagePlatforms(ctx, img, imgID, options.Platforms)
+	}
 	return i.deleteAll(ctx, imgID, all, c, prune)
 }
 
+// deleteImagePlatforms iterates over a slice of platforms and deletes each one for the given image.
+func (i *ImageService) deleteImagePlatforms(ctx context.Context, img *c8dimages.Image, imgID image.ID, platformsToDel []ocispec.Platform) ([]imagetypes.DeleteResponse, error) {
+	var accumulatedResponses []imagetypes.DeleteResponse
+	for _, p := range platformsToDel {
+		responses, err := i.deleteImagePlatformByImageID(ctx, img, imgID, &p)
+		if err != nil {
+			return nil, fmt.Errorf("failed to delete platform %s for image %s: %w", platforms.Format(p), imgID.String(), err)
+		}
+		accumulatedResponses = append(accumulatedResponses, responses...)
+	}
+	return accumulatedResponses, nil
+}
+
+func (i *ImageService) deleteImagePlatformByImageID(ctx context.Context, img *c8dimages.Image, imgID image.ID, platform *ocispec.Platform) ([]imagetypes.DeleteResponse, error) {
+	pm := platforms.OnlyStrict(*platform)
+	var target ocispec.Descriptor
+	if img == nil {
+		// Find any image with the same target
+		// We're deleting by digest anyway so it doesn't matter - we just
+		// need a c8d image object to pass to getBestPresentImageManifest
+		i, err := i.resolveImage(ctx, imgID.String())
+		if err != nil {
+			return nil, err
+		}
+		img = &i
+	}
+	imgMfst, err := i.getBestPresentImageManifest(ctx, *img, pm)
+	if err != nil {
+		return nil, err
+	}
+	target = imgMfst.Target()
+
+	var toDelete []ocispec.Descriptor
+	err = i.walkPresentChildren(ctx, target, func(ctx context.Context, d ocispec.Descriptor) error {
+		toDelete = append(toDelete, d)
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Check if these are not used by other images with different
+	// target root images.
+	// The same manifest can be referenced by different image indexes.
+	var response []imagetypes.DeleteResponse
+	for _, d := range toDelete {
+		if err := i.content.Delete(ctx, d.Digest); err != nil {
```

Comment:
We should avoid doing any direct deletion like this, removing root objects and updating reference labels is the preferred solution. There is no other transactional way to ensure nothing is using the content

Reply:
Right, it's not really ideal, but overall we assume that `moby` containerd namespace is exclusively accesses by `dockerd`.

There's a check that at least makes sure that it's not used by any container and that requires an explicit `force`.

Perhaps we should make it ONLY work with `force` now? That would make it clear that the operation is potentially dangerous, while still giving an escape hatch for users that want to get rid of platforms they don't need.